### ISSUE TITLE
Passing whitespace characters to White() not in whiteStrs causes KeyError

### ISSUE THREAD

**Issue by rtkjbillo** [pyparsing/pyparsing#51]
Environment: macOS 10.14.1, Python 2.7.15 from MacPorts, pyparsing 2.3.0 from pypi

When constructing a White() token with characters in the `ws=` parameter that are not in the `White.whiteStrs` dict, such as `\u00a0` (Unicode NO-BREAK SPACE), the constructor raises a KeyError when it tries to set the `self.name` property. Example:

```
Python 2.7.15 (default, Oct  1 2018, 15:59:56)
[GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyparsing import White
>>> w = White(ws=u' \t\u00a0')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pyparsing.py", line 3132, in __init__
    self.name = ("".join(White.whiteStrs[c] for c in self.matchWhite))
  File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/pyparsing.py", line 3132, in <genexpr>
    self.name = ("".join(White.whiteStrs[c] for c in self.matchWhite))
KeyError: u'\xa0'
```

My workaround for now is to manually add the necessary key/value pairs into the `whiteStrs` dict:

```
>>> from pyparsing import White
>>> White.whiteStrs[u'\u00a0'] = "<NO-BREAK SPACE>"
>>> w = White(u' \t\u00a0')
>>> w.name
'<SPC><TAB><NO-BREAK SPACE>'
```

It may be a good idea to provide a method for setting `whiteStrs` entries, or have the name generation process gracefully handle non-existent keys.

**Comment by ptmcg** [comment]
Good suggestion, I'll fold this in to the current development work. 

In general, using White expressions is a last-resort sort of thing with pyparsing. Are these characters that you would want to add to the default whitespace set, so that pyparsing will ignore them for you (instead of you having to inject White() expressions at multiple places in your parser)? Using [`ParserElement.setDefaultWhitespaceChars()`](https://pyparsing-docs.readthedocs.io/en/latest/pyparsing.html?highlight=setdefaultwhite#pyparsing.ParserElement.setDefaultWhitespaceChars)

**Comment by ptmcg** [comment]
Is it simpler if I just define all the whitespace characters in Unicode?

```
    whiteStrs = {
        ' ' : '<SP>',
        '\t': '<TAB>',
        '\n': '<LF>',
        '\r': '<CR>',
        '\f': '<FF>',
        'u\00A0': '<NBSP>',
        'u\1680': '<OGHAM_SPACE_MARK>',
        'u\180E': '<MONGOLIAN_VOWEL_SEPARATOR>',
        'u\2000': '<EN_QUAD>',
        'u\2001': '<EM_QUAD>',
        'u\2002': '<EN_SPACE>',
        'u\2003': '<EM_SPACE>',
        'u\2004': '<THREE-PER-EM_SPACE>',
        'u\2005': '<FOUR-PER-EM_SPACE>',
        'u\2006': '<SIX-PER-EM_SPACE>',
        'u\2007': '<FIGURE_SPACE>',
        'u\2008': '<PUNCTUATION_SPACE>',
        'u\2009': '<THIN_SPACE>',
        'u\200A': '<HAIR_SPACE>',
        'u\200B': '<ZERO_WIDTH_SPACE>',
        'u\202F': '<NNBSP>',
        'u\205F': '<MMSP>',
        'u\3000': '<IDEOGRAPHIC_SPACE>',
        }

```

**Comment by rtkjbillo** [comment]
> Is it simpler if I just define all the whitespace characters in Unicode?

Essentially this is what I've done, with the addition of `u\feff`(ZERO WIDTH NO-BREAK SPACE) - that one is debatable though since Python doesn't seem to evaluate it as True using `isspace()`.

> Are these characters that you would want to add to the default whitespace set, so that pyparsing will ignore them for you (instead of you having to inject White() expressions at multiple places in your parser)?

Yes, I would like them in the default set. I think I still need to use White() tokens - it's entirely possible that my use of White() is not optimal though. 

The background is that I'm writing a query string parser with operators (AND, OR, NOT) using `infixNotation`. The current design is such that "aa bb AND cc" would treat "aa bb" as a single term, with the space included. As a result, I have a base expression and infix parser that are generated like:

```
unicode_chars_nospace = '...' # actually contains the set of all Unicode characters, minus anything evaluating as .isspace()
whitespace_chars = ' \t'  # plus Unicode whitespaces mentioned, but don't need line endings
ParserElement.setDefaultWhitespaceChars(whitespace_chars)
one_char = Char(unicode_chars_nospace)
one_word = Word(unicode_chars_nospace)
leading_space_non_operator_word = White(ws=whitespace_chars) + NotAny(Or(['AND', 'OR', 'NOT'])) + one_word

# One non-space character, followed by one or more non-space characters, followed by 0 or more elements
# that begin with one or more spaces, contain no operators, followed by one or more characters.
# This matches the following terms:
# "aa"
# "aaaaaaaa"
# "aa bb"
# "aa   threespaces"
# "aa 1 2 3"
base_expr = Combine(one_char + one_word + ZeroOrMore(leading_space_non_operator_word))

# We also want to handle the case where we start with one character, followed by one or elements that begin
# with one or more spaces, contain no operators, and are followed by one or more characters.
# This matches the following terms:
# "1 2"
# "3   5"
# "a term"
# "a longer term"
# "a     fivespaces"
base_expr |= Combine(one_char + OneOrMore(leading_space_non_operator_word))
base_expr.leaveWhitespace()

infix_parser = infixNotation(base_expr, [
  ('NOT', 1, opAssoc.RIGHT, NotOperator),
  ('AND', 2, opAssoc.LEFT, AndOperator),
  ('OR', 2, opAssoc.LEFT, OrOperator)
])
infix_parser.leaveWhitespace()
```

Perhaps there is a better way to do this, but my intent is to preserve the number and position of spaces in a single term - for example, `Error on interface AND eth0` should have `Error on interface` as a single term. Another example would be `[INFO]    Message is padded OR [WARNING]`, where the spaces after `[INFO]    ` are relevant and included in the term.

**Comment by ptmcg** [comment]
Here is an alternative that avoids the `White()` issues altogether. Instead I use a combination of `OneOrMore`, which does whitespace skipping based on the `setDefaultWhitespaceChars` value; wrapped in `originalTextFor`, which gives back the original substring from the input, whitespace and all.

```
whitespace_chars = ' \t'  # plus Unicode whitespaces mentioned, but don't need line endings
ParserElement.setDefaultWhitespaceChars(whitespace_chars)

# replace with unicode set of your choice, or just pyparsing_unicode.printables, 
# though this will be slow
unicode_printable_chars = pyparsing_unicode.Latin1.printables

# use keyword classes instead of literals, to avoid accidental match with leading 
# substring in a term, such as NOTHING, ANDROID, OREGON, etc.
NOT, AND, OR = map(CaselessKeyword, "NOT AND OR".split())
base_word = ~(NOT | AND | OR) + Word(unicode_printable_chars)

base_expr = originalTextFor(OneOrMore(base_word))

infix_parser = infixNotation(base_expr, [
  (NOT, 1, opAssoc.RIGHT, ), # NotOperator),
  (AND, 2, opAssoc.LEFT, ), #AndOperator),
  (OR, 2, opAssoc.LEFT, ), #OrOperator)
])

infix_parser.runTests("""\
aa
aaaaaaaa
aa bb
aa   threespaces
aa 1 2 3
1 2
3   5
a term
a longer term
a     fivespaces
a     fivespaces or 3  5
Error on interface AND eth0
[INFO] Message is padded OR [WARNING]
""")

```
Prints:
```
aa
['aa']


aaaaaaaa
['aaaaaaaa']


aa bb
['aa bb']


aa   threespaces
['aa   threespaces']


aa 1 2 3
['aa 1 2 3']


1 2
['1 2']


3   5
['3   5']


a term
['a term']


a longer term
['a longer term']


a     fivespaces
['a     fivespaces']


a     fivespaces or 3  5
[['a     fivespaces', 'OR', '3  5']]
[0]:
  ['a     fivespaces', 'OR', '3  5']


Error on interface AND eth0
[['Error on interface', 'AND', 'eth0']]
[0]:
  ['Error on interface', 'AND', 'eth0']


[INFO] Message is padded OR [WARNING]
[['[INFO] Message is padded', 'OR', '[WARNING]']]
[0]:
  ['[INFO] Message is padded', 'OR', '[WARNING]']
```

But I still like the addition of new unicode whitespace characters in the `White()` class character map, and will include them in the next release.

**Comment by rtkjbillo** [comment]
Just getting back to this - thanks very much for the example! I'm going to benchmark it against what we have now, which also includes a pre-parser to pre-emptively reject unbalanced parentheses and other potentially invalid combinations of terms and operators.

**Comment by ptmcg** [comment]
Released in 2.3.1

**Comment by eboth** [comment]
I'm still getting the same KeyError in 2.4.5 on Python 3.7.2. As detailed in Jake's workaround, the error goes away if I add `'\u00a0'` to whiteStrs manually.

Looking at `core.py`, it seems that the unicode entities are simply mistyped (the backslash and the u reversed, e.g. `'u\00a0'` instead of `'\u00a0'`), as indeed seems to be the case in the [list posted in this thread](https://github.com/pyparsing/pyparsing/issues/51#issuecomment-442471464-permalink).
